File: test/features/ocr/phone_extractor_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:war2aty/features/ocr/domain/services/phone_extractor.dart';

void main() {
  const extractor = PhoneExtractor();

  group('PhoneExtractor', () {
    group('local mobile numbers', () {
      test('contiguous 11-digit mobile', () {
        final results = extractor.extract('اتصل 01012345678');
        expect(results, hasLength(1));
        expect(results.first.normalizedNumber, equals('01012345678'));
        expect(results.first.isAmbiguous, isFalse);
      });

      test('010 prefix', () {
        final results = extractor.extract('01098765432');
        expect(results, hasLength(1));
      });

      test('011 prefix', () {
        final results = extractor.extract('01112345678');
        expect(results, hasLength(1));
      });

      test('012 prefix', () {
        final results = extractor.extract('01212345678');
        expect(results, hasLength(1));
      });

      test('015 prefix', () {
        final results = extractor.extract('01512345678');
        expect(results, hasLength(1));
      });
    });

    group('whitespace reassembly', () {
      test('spaces between groups', () {
        final results = extractor.extract('010 1234 5678');
        expect(results, hasLength(1));
        expect(results.first.normalizedNumber, equals('01012345678'));
        expect(results.first.isAmbiguous, isTrue);
      });

      test('dashes between groups', () {
        final results = extractor.extract('010-1234-5678');
        expect(results, hasLength(1));
        expect(results.first.normalizedNumber, equals('01012345678'));
      });

      test('dots between groups', () {
        final results = extractor.extract('010.1234.5678');
        expect(results, hasLength(1));
        expect(results.first.normalizedNumber, equals('01012345678'));
      });

      test('mixed separators', () {
        final results = extractor.extract('010 1234-5678');
        expect(results, hasLength(1));
        expect(results.first.normalizedNumber, equals('01012345678'));
      });
    });

    group('international format', () {
      test('+20 prefix', () {
        final results = extractor.extract('+201012345678');
        expect(results, hasLength(1));
        expect(results.first.normalizedNumber, equals('01012345678'));
      });

      test('002 prefix', () {
        final results = extractor.extract('00201012345678');
        expect(results, hasLength(1));
        expect(results.first.normalizedNumber, equals('01012345678'));
      });
    });

    group('invalid numbers', () {
      test('rejects non-Egyptian prefix', () {
        final results = extractor.extract('01312345678');
        expect(results, isEmpty);
      });

      test('rejects too few digits', () {
        final results = extractor.extract('0101234567');
        expect(results, isEmpty);
      });

      test('deduplicates same number appearing twice', () {
        final results = extractor.extract('01012345678 or 01012345678');
        expect(results, hasLength(1));
      });
    });

    group('multiple numbers', () {
      test('extracts two different numbers', () {
        final results = extractor.extract(
          'خدمة العملاء 01012345678 أو 01198765432',
        );
        expect(results, hasLength(2));
      });
    });
  });
}

File: supabase/tests/unit/phone-extractor.test.ts
/**
 * F13-T05 · Tests for the phone extractor.
 *
 * Mirrors `test/features/ocr/phone_extractor_test.dart` case-for-case.
 */

import { assertEquals } from "jsr:@std/assert@1";

import { extractPhones } from "../../functions/_shared/extractors/phone-extractor.ts";

// ── local mobile numbers ─────────────────────────────────────────────────────

Deno.test("phone: contiguous 11-digit mobile", () => {
  const results = extractPhones("اتصل 01012345678");
  assertEquals(results.length, 1);
  assertEquals(results[0].normalized_number, "01012345678");
  assertEquals(results[0].is_ambiguous, false);
});

Deno.test("phone: 010 prefix", () => {
  assertEquals(extractPhones("01098765432").length, 1);
});

Deno.test("phone: 011 prefix", () => {
  assertEquals(extractPhones("01112345678").length, 1);
});

Deno.test("phone: 012 prefix", () => {
  assertEquals(extractPhones("01212345678").length, 1);
});

Deno.test("phone: 015 prefix", () => {
  assertEquals(extractPhones("01512345678").length, 1);
});

// ── whitespace reassembly ────────────────────────────────────────────────────

Deno.test("phone: spaces between groups", () => {
  const results = extractPhones("010 1234 5678");
  assertEquals(results.length, 1);
  assertEquals(results[0].normalized_number, "01012345678");
  assertEquals(results[0].is_ambiguous, true);
});

Deno.test("phone: dashes between groups", () => {
  const results = extractPhones("010-1234-5678");
  assertEquals(results.length, 1);
  assertEquals(results[0].normalized_number, "01012345678");
});

Deno.test("phone: dots between groups", () => {
  const results = extractPhones("010.1234.5678");
  assertEquals(results.length, 1);
  assertEquals(results[0].normalized_number, "01012345678");
});

Deno.test("phone: mixed separators", () => {
  const results = extractPhones("010 1234-5678");
  assertEquals(results.length, 1);
  assertEquals(results[0].normalized_number, "01012345678");
});

// ── international format ─────────────────────────────────────────────────────

Deno.test("phone: +20 prefix", () => {
  const results = extractPhones("+201012345678");
  assertEquals(results.length, 1);
  assertEquals(results[0].normalized_number, "01012345678");
});

Deno.test("phone: 002 prefix", () => {
  const results = extractPhones("00201012345678");
  assertEquals(results.length, 1);
  assertEquals(results[0].normalized_number, "01012345678");
});

// ── invalid numbers ──────────────────────────────────────────────────────────

Deno.test("phone: rejects non-Egyptian prefix", () => {
  assertEquals(extractPhones("01312345678"), []);
});

Deno.test("phone: rejects too few digits", () => {
  assertEquals(extractPhones("0101234567"), []);
});

Deno.test("phone: deduplicates same number appearing twice", () => {
  assertEquals(extractPhones("01012345678 or 01012345678").length, 1);
});

// ── multiple numbers ─────────────────────────────────────────────────────────

Deno.test("phone: extracts two different numbers", () => {
  const results = extractPhones("خدمة العملاء 01012345678 أو 01198765432");
  assertEquals(results.length, 2);
});
